### PR TITLE
sfapi: direct-SFAPI (NERSC) launch path as a standalone sfapi_connect plugin

### PR DESCRIPTION
**Pull request by andre-merzky**
## Motivation

NERSC's IRI facility API (`api.iri.nersc.gov`) is currently broken server-side: every `/api/v1/compute/*` route fails with a wrapped internal SFAPI 500 (`Unable to get sfapi data: 500`) while account/status routes work — verified with a fresh, working SFAPI client, and reported to the IRI operators. This blocks all orbit endpoint launches on Perlmutter through the IRI path.

This PR adds a direct-SFAPI launch path at feature parity with the IRI plugin, as a standalone broker plugin.

## What's included

**New plugin pair**
- `plugin_sfapi_connect.py` — broker-hosted `sfapi_connect`: `connect(endpoint, client_id, private_key)` (OAuth2 client-credentials + `private_key_jwt` via `authlib` against `oidc.nersc.gov`), dynamic per-endpoint instances named `sfapi.<endpoint>`. Credentials live in broker memory only. Eager token mint **before** instance registration (bad creds → 401, never a half-registered instance); reconnect verifies new credentials before rotating; missing `authlib` → 501.
- `plugin_sfapi_instance.py` — same route surface as the IRI instance plugin (Explorer-compatible): submit/status/list/cancel via SFAPI's task-mediated flow (`POST /compute/jobs/{machine}` form-encoded → task poll → jobid), sacct-backed status polling emitting the standard `job_status` notifications, sbatch-script renderer from PSiJ-flavored job specs, token auto-refresh with single-flight locking and 401-retry-once.
- `sfapi_endpoints.py` registry; read-only Explorer page (deliberately no key-paste form — private keys must not transit the browser).

**Launch-contract fixes** (apply to all channels, found by live debugging on Perlmutter)
- Wrapper exports `PATH="@BINDIR@:$PATH"` — dragon resolves its WLM helpers (`dragon-network-config-launch-helper`, `dragon-backend`) *by name* through srun; absolute-path anchoring of dragon/python alone is not enough.
- Wrapper forces `SLURM_EXPORT_ENV=ALL` under Slurm — `--export=NONE` submissions otherwise scrub the env from inner job steps, losing that PATH.
- The sbatch renderer emits `#SBATCH --export=NONE`: SFAPI-submitted jobs inherit the user's full captured login environment, which silently wedges dragon's backend bring-up; all required variables travel as explicit `export` lines in the script body, making the job env deterministic by construction.
- Wrapper comment now warns that dragon 0.14's channel-based logging (`-l dragon_file=DEBUG -l stderr=DEBUG`) deadlocks backend bring-up right after `BEIsUp` (verified live) — post-mortem use only.

**Cleanups**
- `httpx`/`httpcore` loggers pinned to WARNING (status polling drowned the broker log in per-request INFO lines).
- `_iri_raise` (earlier commit on devel) passes upstream diagnostics through; the sfapi error mapping reuses it.

**Docs** — DEPLOYMENT.md gains an "SFAPI (NERSC) launch path" section: client creation at iris.nersc.gov (IP ranges fixed at creation, /24 max width, Spin preset for broker-held credentials; 2/30/60-day client expiry surfacing as auth failures), credential handling, and the deterministic-environment design.

## Validation

- Unit suite: **963 passed** (46 new tests: renderer incl. `--export=NONE`, token manager refresh/single-flight/error mapping, task-poll edge cases incl. completed-without-jobid and poll-timeout-with-task-id, connect dispatch incl. mint-failure-never-registers, script-never-logged caplog guard). flake8 clean.
- The design went through two adversarial review rounds before implementation.
- **Live end-to-end on Perlmutter (2026-07-11)**: SFAPI connect → sbatch submission → endpoint job → dragon bring-up → forward tunnel → broker registration → ROSE active-learning workload (rhapsody function tasks on the compute node) → teardown.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/radical/orbit/plugin_sfapi_connect.py [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+
+        An access token is minted from the given credentials *before* the
+        instance is registered / rotated, so bad credentials fail fast with
+        401 and never leave a half-registered instance behind (no deregister
+        path, no register/deregister race with a concurrent reconnect).
+        '''
+        try:
+            data = await request.json()
+        except Exception:
+            data = {}
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/orbit/plugin_sfapi_connect.py:217]: Fixed in d9a5450 — the reply shape is now validated before field access (plus test coverage for the non-object submit reply and scalar list replies).

File: src/radical/orbit/plugin_sfapi_instance.py [hunk 2]
```diff
@@ -0,0 +1,834 @@
...
+                'POST', f'/compute/jobs/{resource_id}',
+                data={'job': script, 'isPath': 'false'})
+        except httpx.RequestError as exc:
+            raise HTTPException(status_code=502,
+                                detail=f'SFAPI submit_job: {exc}') from exc
+
+        _sfapi_raise(resp, 'submit_job')
+
+        task          = resp.json()
+        task_id       = str(task.get('id') or task.get('task_id') or '')
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/orbit/plugin_sfapi_instance.py]: Fixed in d9a5450 — the reply shape is now validated before field access (plus test coverage for the non-object submit reply and scalar list replies).

File: src/radical/orbit/plugin_sfapi_instance.py [hunk 3]
```diff
@@ -0,0 +1,834 @@
...
+        self._check_active()
+        try:
+            resp = await self._request('GET', '/status')
+        except httpx.RequestError as exc:
+            raise HTTPException(status_code=502,
+                                detail=f'SFAPI list_resources: {exc}') from exc
+        _sfapi_raise(resp, 'list_resources')
+        data      = resp.json()
+        resources = data if isinstance(data, list) \
+                         else data.get('resources', data.get('systems', []))
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/orbit/plugin_sfapi_instance.py]: Fixed in d9a5450 — the reply shape is now validated before field access (plus test coverage for the non-object submit reply and scalar list replies).

File: src/radical/orbit/plugin_sfapi_instance.py [hunk 4]
```diff
@@ -0,0 +1,834 @@
...
+        self._check_active()
+        try:
+            resp = await self._request('GET', '/status/outages')
+        except httpx.RequestError as exc:
+            raise HTTPException(status_code=502,
+                                detail=f'SFAPI list_incidents: {exc}') from exc
+        _sfapi_raise(resp, 'list_incidents')
+        data      = resp.json()
+        incidents = data if isinstance(data, list) \
+                         else data.get('outages', data.get('incidents', data))
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/orbit/plugin_sfapi_instance.py]: Fixed in d9a5450 — the reply shape is now validated before field access (plus test coverage for the non-object submit reply and scalar list replies).

File: src/radical/orbit/plugin_sfapi_instance.py [hunk 5]
```diff
@@ -0,0 +1,834 @@
...
+        self._check_active()
+        try:
+            resp = await self._request('GET', '/account/projects')
+        except httpx.RequestError as exc:
+            raise HTTPException(status_code=502,
+                                detail=f'SFAPI list_projects: {exc}') from exc
+        _sfapi_raise(resp, 'list_projects')
+        data     = resp.json()
+        projects = data if isinstance(data, list) \
+                        else data.get('projects', data)
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/orbit/plugin_sfapi_instance.py]: Fixed in d9a5450 — the reply shape is now validated before field access (plus test coverage for the non-object submit reply and scalar list replies).

**Comment by andre-merzky** [comment]
Pushed 2b66eca — fixes the job-status poller blindness found while investigating why an externally-cancelled job never tripped the client's failure watch.

TL;DR: `GET /compute/jobs/{machine}/{jobid}?sacct=true` is cache-backed at NERSC and chronically returns empty `output` — the poller never saw any state for any job (masked by the deliberate empty-output tolerance for the post-submit sacct lag). NERSC's own `sfapi_client` avoids that route entirely; the poller and `get_job_status` now use its recipe: `GET /compute/jobs/{machine}?sacct=true&cached=false&kwargs=jobid=N` (live-verified against the incident job: returns `CANCELLED by …`). Parent-row selection guards against `.batch`/`.extern` step rows. The notification delivery chain itself (hosted plugin → EventRouter → subscriber) was verified sound with a transport-level reproduction.

966 tests green, flake8 clean.
